File: parking-web-common/src/main/java/com/chatchatabc/parking/web/common/impl/application/rest/service/JwtServiceImpl.java
package com.chatchatabc.parking.web.common.impl.application.rest.service;

import com.auth0.jwt.JWT;
import com.auth0.jwt.JWTVerifier;
import com.auth0.jwt.algorithms.Algorithm;
import com.auth0.jwt.interfaces.Payload;
import com.chatchatabc.parking.web.common.application.rest.service.JwtService;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.Date;
import java.util.List;
import java.util.Map;

@Service
public class JwtServiceImpl implements JwtService {
    private final Long expiration;
    private final ObjectMapper objectMapper;


    private final Algorithm hmac512;
    private final JWTVerifier verifier;

    public JwtServiceImpl(
            @Value("${server.jwt.secret}")
            String secret,
            @Value("${server.jwt.expiration}")
            Long expiration,
            ObjectMapper objectMapper
    ) {
        this.objectMapper = objectMapper;
        this.expiration = expiration;
        hmac512 = Algorithm.HMAC512(secret);
        verifier = JWT.require(hmac512).build();
    }

    /**
     * Generate a JWT token for the given user id
     *
     * @param userId   the user id
     * @param username the username
     * @param roles    the roles
     * @return the JWT token
     */
    @Override
    public String generateToken(String userId, String username, List<String> roles) {
        return JWT.create()
                .withSubject(userId)
                .withIssuer("DavaoParking")
                .withClaim("username", username)
                .withClaim("role", roles)
                .withIssuedAt(new Date(System.currentTimeMillis()))
                .withExpiresAt(new Date(System.currentTimeMillis() + expiration))
                .sign(hmac512);
    }

    /**
     * Validate the given token and return the jwt payload
     *
     * @param token the token
     * @return the jwt payload
     */
    @Override
    public Payload validateTokenAndGetPayload(String token) {
        try {
            return verifier.verify(token);
        } catch (Exception e) {
            return null;
        }
    }

    /**
     * Get the expiration from the given token
     *
     * @param token the token
     * @return the expiration
     */
    @Override
    public Long getExpirationFromToken(String token) {
        try {
            // Split the token into header, payload, and signature
            String[] parts = token.split("\\.");
            if (parts.length < 2) {
                return null; // Invalid token
            }

            // Decode the payload
            String payload = parts[1];
            String decodedPayload = new String(Base64.getUrlDecoder().decode(payload), StandardCharsets.UTF_8);

            // Parse the payload as JSON and retrieve the "exp" claim
            Map<String, Object> claims = objectMapper.readValue(decodedPayload, Map.class);
            if (claims.containsKey("exp")) {
                Number expNumber = (Number) claims.get("exp");
                return expNumber.longValue();
            }
        } catch (Exception ignored) {
        }
        return null;
    }
}

File: backend/parking-service/src/main/kotlin/com/chatchatabc/service/impl/application/rest/service/JwtServiceImpl.kt
package com.chatchatabc.service.impl.application.rest.service

import com.auth0.jwt.JWT
import com.auth0.jwt.JWTVerifier
import com.auth0.jwt.algorithms.Algorithm
import com.chatchatabc.api.application.dto.user.UserDTO
import com.chatchatabc.api.application.rest.service.JwtService
import com.chatchatabc.service.domain.repository.UserRepository
import org.apache.dubbo.config.annotation.DubboService
import org.modelmapper.ModelMapper
import org.springframework.beans.factory.annotation.Value
import java.util.*

@DubboService
class JwtServiceImpl(
    @Value("\${server.jwt.secret}")
    private val secret: String,

    @Value("\${server.jwt.expiration}")
    private val expiration: String,

    private val userRepository: UserRepository
) : JwtService {
    private val hmac512: Algorithm = Algorithm.HMAC512(secret)
    private val verifier: JWTVerifier = JWT.require(hmac512).build()
    private val modelMapper = ModelMapper()

    /**
     * Generate Token
     */
    override fun generateToken(user: UserDTO): String? {
        return JWT.create()
            .withSubject(user.id)
            .withExpiresAt(Date(System.currentTimeMillis() + expiration.toLong()))
            .sign(hmac512)
    }

    /**
     * Validate Token and Get User
     */
    override fun validateTokenAndGetUser(token: String): UserDTO? {
        return try {
            val subject = verifier.verify(token).subject
            val user = userRepository.findById(subject)
            if (user.isEmpty) {
                throw Exception("Invalid Token")
            }
            modelMapper.map(user.get(), UserDTO::class.java)
        } catch (e: Exception) {
            return null
        }
    }
}